pentasharp/Controllers/AuthenticateController.cs

```diff
@@ -128,5 +130,39 @@ public IActionResult DeleteConfirmed(int id)
             }
             return RedirectToAction("UserList");
         }
+
+        public IActionResult Login()
+        {
+            return View(new LoginViewModel());
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Login(LoginViewModel model)
+        {
+            if (!ModelState.IsValid)
+            {
+                return View(model);
+            }
+
+            var user = await _context.Users.SingleOrDefaultAsync(u => u.Email == model.Email);
+            if (user == null || user.PasswordHash != HashPassword(model.Password))
+            {
+                ModelState.AddModelError(string.Empty, "Email or password are incorrect.");
+                return View(model);
+            }
+
+            HttpContext.Session.SetString("UserId", user.UserId.ToString());
```

Comment:
Use dependency injection to access HttpContext and manage session instead of directly calling HttpContext.